### ISSUE TITLE
Upgrading Salt RPM stops minion service

### ISSUE THREAD

**Issue by cedwards** [saltstack/salt#28424]
I discovered during a recent upgrade to 2015.8.1 (from 2015.5.5) that the minion service stopped after upgrading the package. This makes a large scale upgrade a huge pain if not upgrading using a state including an `at` job to restart the service.

@terminalmage, @dmurphy18 ? Any thoughts here?

**Comment by cachedout** [comment]
Yikes! Yes, that would be bad. I'll see if I can find somebody to look at this ASAP.

**Comment by terminalmage** [comment]
We don't have anything in the post-install steps for the RPM spec which will start the service. We do have pre-uninstall steps which stop the salt-minion service, however, which is leading to what you're seeing.

There is no distinction in an RPM spec file, to my knowledge, for post-install actions that happen only when upgrading, so as I as see it we have two options:

1) Add options to ``%post`` in the RPM spec, which means that all new installs will automatically start the minion service after the package is installed (I don't like this.)

2) Manage the service's state using Salt:

```
salt-minion:
  service:
    - running
    - watch:
      - pkg: salt-minion
  pkg:
    - installed
```

**Comment by lorengordon** [comment]
Maybe use %pre to check whether the services are enabled or running, and %post to attempt to restore that same state? Here's a link with a section on saving state between rpm scriptlets...

https://fedoraproject.org/wiki/Packaging:ScriptletSnippets

**Comment by variia** [comment]
@lorengordon the lock file is meant to do the same, no need to do anything in %pre or %post.

@terminalmage is right, #1 is a no no, #2 is vital but only if the service survives upgrade and of course running.
I only looked at the 2015.5.5 spec file but I don't see anything wrong it. The %preuninstall script should not trigger 'stop' in case of upgrade and %postuninstall should bounce the service in case lock file exist.

https://docs.fedoraproject.org/en-US/Fedora_Draft_Documentation/0.1/html/RPM_Guide/ch09s04s05.html

**Comment by stardust85** [comment]
Can his be related to https://github.com/saltstack/salt/issues/28474?

**Comment by variia** [comment]
@stardust85 hmm, not sure. I have just spent significant amount of time trying to find out what is causing this. I have upgraded from 2015.5.6 to 2015.8.1 from SaltStack repo and first time I had these upgrading:
```
Nov 06 17:42:18 Updated: PyYAML-3.11-1.el6.x86_64
Nov 06 17:42:18 Updated: python-crypto-2.6.1-2.el6.x86_64
Nov 06 17:42:19 Installed: python-markupsafe-0.11-10.el6.x86_64
Nov 06 17:42:19 Installed: zeromq-4.0.5-1.el6.x86_64
Nov 06 17:42:20 Updated: python-zmq-14.5.0-2.el6.x86_64
Nov 06 17:42:20 Installed: python-tornado-4.2.1-1.el6.x86_64
Nov 06 17:42:20 Installed: python-futures-3.0.3-1.el6.noarch
Nov 06 17:42:24 Updated: salt-2015.8.1-1.el6.noarch
Nov 06 17:42:25 Updated: salt-minion-2015.8.1-1.el6.noarch
Nov 06 17:42:25 Updated: openpgm-5.2.122-2.el6.x86_64
Nov 06 17:42:26 Erased: zeromq3
```

And my minion was purely shut down during uninstall and it stopped there:
```
2015-11-06 17:36:10,209 [salt.loaded.int.module.saltutil          ][INFO    ][1127] Creating module dir '/var/cache/salt/minion/extmods/utils'
2015-11-06 17:36:10,209 [salt.loaded.int.module.saltutil          ][INFO    ][1127] Syncing utils for environment 'base'
2015-11-06 17:36:10,209 [salt.loaded.int.module.saltutil          ][INFO    ][1127] Loading cache from salt://_utils, for base)
2015-11-06 17:36:10,209 [salt.fileclient                          ][INFO    ][1127] Caching directory '_utils/' for environment 'base'
2015-11-06 17:36:10,252 [salt.minion                              ][INFO    ][1127] Returning information for job: 20151106173608747427
2015-11-06 17:36:11,101 [salt.minion                              ][INFO    ][1130] Returning information for job: 20151106173608750363
2015-11-06 17:42:25,560 [salt.cli.daemons                         ][INFO    ][1054] The salt minion is shut down
```

Now, what I have done next is erased my salt and salt-minion rpms, reinstalled 2015.5.6 (leaving all extra python/zmq dependencies that got installed as part of 2015.8.1) then upgraded again to 2015.8.1. Second time, I only had salt upgrades and I cannot reproduce this issue, the minion gets bounced nicely with condrestart as per rpm design:
```
2015-11-06 18:49:13,219 [salt.cli.daemons ][INFO    ][10506] Setting up the Salt Minion "salt-centos-dev.eu-west-1.compute.internal"
2015-11-06 18:49:13,517 [salt.cli.daemons                         ][INFO    ][10506] The salt minion is shut down
2015-11-06 18:49:13,547 [salt.cli.daemons                         ][INFO    ][10512] The salt minion is shut down
2015-11-06 18:49:14,808 [salt.cli.daemons                         ][INFO    ][10514] The salt minion is starting up
2015-11-06 18:49:14,809 [salt.minion                              ][INFO    ][10514] Minion is starting as user 'root'
2015-11-06 18:49:14,812 [salt.utils.event                         ][INFO    ][10514] Starting pub socket on ipc:///var/run/salt/minion/minion_event_dd9a8b0fc1_pub.ipc
2015-11-06 18:49:14,813 [salt.utils.event                         ][INFO    ][10514] Starting pull socket on ipc:///var/run/salt/minion/minion_event_dd9a8b0fc1_pull.ipc
2015-11-06 18:49:17,861 [salt.loaded.int.module.cmdmod            ][INFO    ][10514] Executing command ['date', '+%z'] in directory '/root'
2015-11-06 18:49:17,907 [salt.minion                              ][INFO    ][10514] Added mine.update to scheduler
2015-11-06 18:49:17,907 [salt.utils.schedule                      ][INFO    ][10514] Updating job settings for scheduled job: __mine_interval
```

And finally, I erased all packages again including all associated python, YAML, zmq packages and reinstalled 2015.5.5 again (from EPEL, disabled SaltStack repo) followed by upgrade to 2015.8.1.

I can confirm that this issue seems to be associated with the extra dependencies picked up and installed from SaltStack repo.

**Comment by dmurphy18** [comment]
Also see related #22973

**Comment by terminalmage** [comment]
@lorengordon @cedwards After doing some more investigation into this, our scriptlet setup is fine. The ``"$1"`` check in the scriptlet controls whether or not the ``condrestart`` is performed, and the fact that we're running a ``condrestart`` means that if the service wasn't already running, it won't reattempt the restart.

Furthermore, I checked the RPMs for 2015.8.3 and 2015.5.5 (as well as a build of 2015.5.8 that I've been working on) and their scriptlets all match.

This seems to be related to python-crypto being updated during the install process. If I manually upgrade python-crypto first, the services are successfully restarted when the salt packages are upgraded. Conversely, I can install all other upgrades/new packages needed by 2015.8.3 *except* for python-crypto, and the salt-master/salt-minion daemons will fail to restart after the install.

I've investigated using the python-crypto2.6 package from EPEL, but it is poorly-made and unusable.

The proper fix for this, as I see it, is as follows:

1. Make a working pycrypto 2.6 package which can be installed alongside python-crypto from EPEL. (I've already begun working on this)
2. Update all imports for pycrypto within the Salt source code to first try the alternate package name (i.e. ``from Crypto26.Cipher import PKCS1_OAEP``, and falling back to ``from Crypto.Cipher import PKCS1_OAEP`` in the case of an ``ImportError``).

Thoughts?
